src/webgpu/api/validation/attachment_compatibility.spec.ts

```diff
@@ -0,0 +1,305 @@
+      loadValue: [0, 0, 0, 0],
+    };
+  }
+
+  createDepthAttachment(format: GPUTextureFormat): GPURenderPassDepthStencilAttachmentDescriptor {
+    return {
+      attachment: this.createAttachmentTextureView(format),
+      depthLoadValue: 0,
+      depthStoreOp: 'clear',
+      stencilLoadValue: 1,
+      stencilStoreOp: 'clear',
+    };
+  }
+
+  createPassOrBundleEncoder(
+    encoderType: 'render pass' | 'render bundle',
+    colorFormats: Iterable<GPUTextureFormat>,
+    depthStencilFormat?: GPUTextureFormat
+  ): CommandBufferMaker<GPURenderPassEncoder | GPURenderBundleEncoder> {
+    const encoder = this.device.createCommandEncoder();
+    const passDesc: GPURenderPassDescriptor = {
+      colorAttachments: Array.from(colorFormats, format => this.createColorAttachment(format)),
+      depthStencilAttachment:
+        depthStencilFormat !== undefined
+          ? this.createDepthAttachment(depthStencilFormat)
+          : undefined,
+    };
+    const pass = encoder.beginRenderPass(passDesc);
+    switch (encoderType) {
+      case 'render bundle': {
+        const bundleEncoder = this.device.createRenderBundleEncoder({
+          colorFormats,
+          depthStencilFormat,
+        });
+
+        return {
+          encoder: bundleEncoder,
+          finish() {
+            const bundle = bundleEncoder.finish();
+            pass.executeBundles([bundle]);
+            pass.endPass();
+            return encoder.finish();
+          },
+        };
+      }
+      case 'render pass':
+        return {
+          encoder: pass,
+          finish() {
+            pass.endPass();
+            return encoder.finish();
+          },
+        };
+    }
+  }
+
+  createRenderPipeline(
+    colorStates: Iterable<GPUColorStateDescriptor>,
+    depthStencilState?: GPUDepthStencilStateDescriptor
+  ) {
+    const wgslVertex = `
+    [[builtin(position)]] var<out> position : vec4<f32>;
+
+    [[stage(vertex)]]
+    fn main() -> void {
+      position = vec4<f32>(0.0, 0.0, 0.0, 0.0);
+      return;
+    }
+  `;
+    const wgslFragment = `
+    [[stage(fragment)]]
+    fn main() -> void {
+      return;
+    }
+  `;
+
+    return this.device.createRenderPipeline({
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'main',
+      },
+      primitiveTopology: 'triangle-list',
+      colorStates,
+      depthStencilState,
+    });
+  }
+}
+
+export const g = makeTestGroup(F);
+
+const kColorAttachmentFormats = kRegularTextureFormats
+  .map((key): [RegularTextureFormat, typeof kRegularTextureFormatInfo[RegularTextureFormat]] => [
+    key,
+    kRegularTextureFormatInfo[key],
+  ])
+  .filter(([, format]) => format.color && format.renderable)
+  .map(([key]) => key);
+
```

Comment:
We should also test that the sampleCount matches.

Reply:
+1, can be a TODO and followup, if you want (because this PR is almost done)